api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java

```diff
@@ -827,21 +815,185 @@ private String generateRandomEightCharacterQualifier() {
-                prerequisites
-                    + "\n\n"
-                    + generateNotebookUserCode(
-                        entry.getValue(),
-                        Domain.fromValue(entry.getKey()),
-                        dataSetExportRequest.getDataSetRequest().getName(),
-                        dbWorkspace.getCdrVersion().getName(),
-                        qualifier,
-                        dataSetExportRequest.getKernelType()))
+    return Stream.concat(
+            queriesByDomain.entrySet().stream()
+                .filter(
+                    query ->
+                        !Domain.WHOLE_GENOME_VARIANT
+                            .toString()
+                            .equals(query.getKey())) // This filter can be removed once valid SQL is
+                // generated for WGS
+                .map(
+                    entry ->
+                        prerequisites
+                            + "\n\n"
+                            + generateNotebookUserCode(
+                                entry.getValue(),
+                                Domain.fromValue(entry.getKey()),
+                                dataSetExportRequest.getDataSetRequest().getName(),
+                                dbWorkspace.getCdrVersion().getName(),
+                                qualifier,
+                                dataSetExportRequest.getKernelType())),
+            generateWgsCode(dataSetExportRequest, dbWorkspace, qualifier).stream())
         .collect(Collectors.toList());
   }
 
+  private List<String> generateWgsCode(
+      DataSetExportRequest dataSetExportRequest, DbWorkspace dbWorkspace, String qualifier) {
+    if (!dataSetExportRequest.getGenerateGenomicsAnalysisCode()) {
+      return new ArrayList<>();
+    }
+
+    if (!workbenchConfigProvider.get().featureFlags.enableGenomicExtraction) {
+      throw new NotImplementedException();
+    }
+
+    if (Strings.isNullOrEmpty(dbWorkspace.getCdrVersion().getWgsBigqueryDataset())) {
+      throw new FailedPreconditionException(
+          "The workspace CDR version does not have whole genome data");
+    }
+
+    if (dataSetExportRequest.getKernelType().equals(KernelTypeEnum.R)) {
+      return generateGenomicsAnalysisComment_R();
+    }
+
+    switch (dataSetExportRequest.getGenomicsAnalysisTool()) {
+      case HAIL:
+        return generateHailCode(qualifier, dataSetExportRequest);
+      case PLINK:
+        return generatePlinkCode(qualifier, dataSetExportRequest);
+      case NONE:
+        return generateDownloadVcfCode(qualifier, dataSetExportRequest);
+      default:
+        throw new BadRequestException("Invalid Genomics Analysis Tool");
+    }
+  }
+
+  private Optional<String> getExtractionDirectory(Long datasetId) {
+    try {
+      return Optional.of(
+              submissionDao
+                  .findMostRecentValidExtractionByDataset(dataSetDao.findById(datasetId).get())
+                  .get()
+                  .getOutputDir()
```

Comment:
If I'm understanding the goal with the try catch, you should instead just be able to do this:

- remove Optional.of
- replace this code with: `.map(DbSubmission::getOutputDir)`
- put replaceFirst into `.map` as well
- remove try/catch

Reply:
so, the `NoSuchElementException` actually handles two optional gets. The first is the `.get()` on `dataSetDao.findById(datasetId).get()` the second is the one on the submissionDao.

Would your pattern work with that constraint? I'm actually not very happy with the way this looks either so I'm happy to try something else. I added quite a few test cases to DataSetServiceTest that covers this function so we should be able to refactor this safely - if you'd like to give it a shot I think it should be easy to pull down and run.

Reply:
Sorry, missed the inner optional. Yeah, this can still work - I don't have enough time left to pull the branch but something like this should work fine:

```
return dataSetDao.findById(datasetId)
  .map(submissionDao::findMostRecentValidExtractionByDataset)
  .map(DbSubumission::getOutputDir)
  .map(dir -> dir.replaceFirst("/$", ""))
  .filter(dir -> !dir.isEmpty());
```